proxy/controller-grpc/src/lib.rs

```diff
@@ -0,0 +1,531 @@
+            } else if reg == Registered::Head.into() {
+                Ok(http::Method::HEAD)
+            } else if reg == Registered::Trace.into() {
+                Ok(http::Method::TRACE)
+            } else {
+                Err(InvalidMethod)
+            },
+            Type::Unregistered(ref m) => {
+                HttpTryFrom::try_from(m.as_str()).map_err(|_| InvalidMethod)
+            }
+        }
+    }
+}
+
+impl<'a> From<&'a http::Method> for common::http_method::Type {
+    fn from(m: &'a http::Method) -> Self {
+        use self::common::http_method::*;
+
+        match *m {
+            http::Method::GET => Type::Registered(Registered::Get.into()),
+            http::Method::POST => Type::Registered(Registered::Post.into()),
+            http::Method::PUT => Type::Registered(Registered::Put.into()),
+            http::Method::DELETE => Type::Registered(Registered::Delete.into()),
+            http::Method::HEAD => Type::Registered(Registered::Head.into()),
+            http::Method::OPTIONS => Type::Registered(Registered::Options.into()),
+            http::Method::CONNECT => Type::Registered(Registered::Connect.into()),
+            http::Method::TRACE => Type::Registered(Registered::Trace.into()),
+            ref method => Type::Unregistered(method.as_str().into()),
+        }
+    }
+}
+
+impl<'a> From<&'a http::Method> for common::HttpMethod {
+    fn from(m: &'a http::Method) -> Self {
+        common::HttpMethod {
+            type_: Some(m.into()),
+        }
+    }
+}
+
+impl<'a> From<&'a http::uri::Scheme> for common::Scheme {
+    fn from(scheme: &'a http::uri::Scheme) -> Self {
+        scheme.as_ref().into()
+    }
+}
+
+impl<'a> From<&'a str> for common::scheme::Type {
+    fn from(s: &'a str) -> Self {
+        use self::common::scheme::*;
+
+        match s {
+            "http" => Type::Registered(Registered::Http.into()),
+            "https" => Type::Registered(Registered::Https.into()),
+            s => Type::Unregistered(s.into()),
+        }
+    }
+}
+
+impl<'a> From<&'a str> for common::Scheme {
+    fn from(s: &'a str) -> Self {
+        common::Scheme {
+            type_: Some(s.into()),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+// TODO: do we want to carry the string if there is one?
+pub struct InvalidMethod;
+
+impl fmt::Display for InvalidMethod {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid http method")
+    }
+}
+
+impl Error for InvalidMethod {
+    #[inline]
+    fn description(&self) -> &str {
+        "invalid http method"
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct InvalidScheme;
+
+impl fmt::Display for InvalidScheme {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "invalid http scheme")
+    }
+}
+
+impl Error for InvalidScheme {
+    #[inline]
+    fn description(&self) -> &str {
+        "invalid http scheme"
+    }
+}
+
+#[cfg(feature = "arbitrary")]
+pub mod arbitrary {
```

Comment:
This is better done in a separate arbitrary.rs file so we don't have 4 extra spaces of indents for each line. If it were tiny it wouldn't matter much, but it's already pretty large.